File: weeks/02/5_counting_valleys/counting_valleys.rb
def countingValleys(steps, path)
    in_mountain = false
    in_valley = false
    mountains = 0
    valleys = 0
    elevation = 0
  
    path.each_char do |p|
      if p == 'D'
        if elevation == 0
          in_valley = true
        end
        elevation -= 1
      end
  
      if p == 'U'
        if elevation == 0
          in_mountain = true
        end
        elevation += 1
      end
  
      if elevation == 0
        if in_mountain
          mountains += 1
        end
        if in_valley
          valleys += 1
        end
        in_mountain = false
        in_valley = false
      end
    end
  
    valleys
  end
  

File: weeks/02/5_counting_valleys/counting_valleys.py
def countingValleys(steps, path):

    inMountain = False
    inValley = False

    mountains = 0
    valleys = 0
    elevation = 0

    for p in path:

        if p == 'D':
            if elevation == 0:
                inValley = True
            elevation -= 1

        if p == 'U':
            if elevation == 0:
                inMountain = True
            elevation += 1

        if elevation == 0:
            if inMountain:
                mountains += 1
            if inValley:
                valleys += 1
            inMountain = False
            inValley = False

    return valleys


print(countingValleys(12, list('DDUUDDUDUUUD')))
